File: backend/Services/InventoryService.cs
using Microsoft.EntityFrameworkCore;
using SnackTracker.Api.Data;
using SnackTracker.Api.Models;

namespace SnackTracker.Api.Services
{
    public class InventoryService
    {
        private readonly SnackTrackerContext _context;

        public InventoryService(SnackTrackerContext context)
        {
            _context = context;
        }

        // Recalculates and updates the price and stock for a snack based on active batches
        public async Task UpdateSnackPriceAndStockAsync(int snackId)
        {
            var snack = await _context.Snacks
                .Include(s => s.Batches)
                .FirstOrDefaultAsync(s => s.SnackId == snackId);

            if (snack == null) return;

            var activeBatches = snack.Batches.Where(b => b.QuantityRemaining > 0).ToList();

            // 1. Update Stock Count
            snack.Stock = activeBatches.Sum(b => b.QuantityRemaining);

            // 2. Calculate Weighted Average Price
            if (activeBatches.Any())
            {
                decimal totalValue = activeBatches.Sum(b => b.QuantityRemaining * b.UnitCost);
                int totalItems = activeBatches.Sum(b => b.QuantityRemaining);

                if (totalItems > 0)
                {
                    decimal rawPrice = totalValue / totalItems;
                    
                    // 3. Apply Rounding Rule (Round UP to nearest 0.05)
                    // Logic: Ceiling(Price / 0.05) * 0.05
                    decimal roundedPrice = Math.Ceiling(rawPrice / 0.05m) * 0.05m;
                    
                    snack.Price = roundedPrice;
                }
            }
            // If no active batches, we retain the last known price.
            
            await _context.SaveChangesAsync();
        }

        // Adds a new inventory batch and updates the price
        public async Task AddBatchAsync(int snackId, int quantity, decimal totalCost)
        {
            if (quantity <= 0) throw new ArgumentException("Quantity must be positive");
            if (totalCost < 0) throw new ArgumentException("Total cost cannot be negative");

            var unitCost = totalCost / quantity;

            var batch = new SnackBatch
            {
                SnackId = snackId,
                QuantityPurchased = quantity,
                QuantityRemaining = quantity,
                UnitCost = unitCost,
                PurchaseDate = DateTime.UtcNow
            };

            _context.SnackBatches.Add(batch);
            await _context.SaveChangesAsync();

            // Trigger recalculation
            await UpdateSnackPriceAndStockAsync(snackId);
        }

        // Consumes stock using FIFO (First-In, First-Out)
        public async Task ConsumeStockAsync(int snackId, int quantityToConsume)
        {
            var snack = await _context.Snacks
                .Include(s => s.Batches)
                .FirstOrDefaultAsync(s => s.SnackId == snackId);
                
            if (snack == null) throw new Exception("Snack not found");
            
            // Verify total stock first
            int totalStock = snack.Batches.Where(b => b.QuantityRemaining > 0).Sum(b => b.QuantityRemaining);
            if (totalStock < quantityToConsume) 
            {
                throw new InvalidOperationException($"Insufficient stock. Requested: {quantityToConsume}, Available: {totalStock}");
            }

            // Get active batches ordered by date (FIFO)
            var batches = snack.Batches
                .Where(b => b.QuantityRemaining > 0)
                .OrderBy(b => b.PurchaseDate)
                .ToList();

            int remainingToConsume = quantityToConsume;

            foreach (var batch in batches)
            {
                if (remainingToConsume <= 0) break;

                if (batch.QuantityRemaining >= remainingToConsume)
                {
                    // This batch has enough to cover the rest
                    batch.QuantityRemaining -= remainingToConsume;
                    remainingToConsume = 0;
                }
                else
                {
                    // Take what we can from this batch
                    remainingToConsume -= batch.QuantityRemaining;
                    batch.QuantityRemaining = 0;
                }
            }

            await _context.SaveChangesAsync();
            
            // Recalculate price (because the ratio of cheap/expensive batches might have changed)
            await UpdateSnackPriceAndStockAsync(snackId);
        }
    }
}

File: backend/src/services/inventoryService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateSnackPriceAndStock = updateSnackPriceAndStock;
exports.addBatch = addBatch;
exports.consumeStock = consumeStock;
// Inventory Service - batch-based inventory with FIFO consumption and weighted-average pricing
const db_1 = require("../db");
/**
 * Recalculate and update the price and stock for a snack based on active batches.
 * Price = weighted average of batch unit costs, rounded UP to nearest $0.05.
 */
function updateSnackPriceAndStock(snackId) {
    const activeBatches = (0, db_1.all)('SELECT * FROM SnackBatches WHERE SnackId = ? AND QuantityRemaining > 0', [snackId]);
    const totalStock = activeBatches.reduce((sum, b) => sum + b.QuantityRemaining, 0);
    if (activeBatches.length > 0 && totalStock > 0) {
        const totalValue = activeBatches.reduce((sum, b) => sum + b.QuantityRemaining * b.UnitCost, 0);
        const rawPrice = totalValue / totalStock;
        // Round UP to nearest $0.05, then fix floating point
        const roundedPrice = Math.ceil(rawPrice / 0.05) * 0.05;
        const finalPrice = Math.round(roundedPrice * 100) / 100;
        (0, db_1.run)('UPDATE Snacks SET Stock = ?, Price = ? WHERE SnackId = ?', [totalStock, finalPrice, snackId]);
    }
    else {
        // No active batches - set stock to 0 but retain last known price
        (0, db_1.run)('UPDATE Snacks SET Stock = ? WHERE SnackId = ?', [totalStock, snackId]);
    }
}
/**
 * Add a new inventory batch and update the snack's price/stock.
 */
function addBatch(snackId, quantity, totalCost) {
    if (quantity <= 0)
        throw new Error('Quantity must be positive');
    if (totalCost < 0)
        throw new Error('Total cost cannot be negative');
    const unitCost = totalCost / quantity;
    const now = new Date().toISOString();
    (0, db_1.run)('INSERT INTO SnackBatches (SnackId, QuantityPurchased, QuantityRemaining, UnitCost, PurchaseDate) VALUES (?, ?, ?, ?, ?)', [snackId, quantity, quantity, unitCost, now]);
    updateSnackPriceAndStock(snackId);
}
/**
 * Consume stock using FIFO (First-In, First-Out).
 * Deducts from the oldest batches first, then recalculates price.
 */
function consumeStock(snackId, quantityToConsume) {
    const batches = (0, db_1.all)('SELECT * FROM SnackBatches WHERE SnackId = ? AND QuantityRemaining > 0 ORDER BY PurchaseDate ASC', [snackId]);
    const totalStock = batches.reduce((sum, b) => sum + b.QuantityRemaining, 0);
    if (totalStock < quantityToConsume) {
        throw new Error(`Insufficient stock. Requested: ${quantityToConsume}, Available: ${totalStock}`);
    }
    let remainingToConsume = quantityToConsume;
    for (const batch of batches) {
        if (remainingToConsume <= 0)
            break;
        const qty = batch.QuantityRemaining;
        if (qty >= remainingToConsume) {
            (0, db_1.run)('UPDATE SnackBatches SET QuantityRemaining = QuantityRemaining - ? WHERE SnackBatchId = ?', [remainingToConsume, batch.SnackBatchId]);
            remainingToConsume = 0;
        }
        else {
            remainingToConsume -= qty;
            (0, db_1.run)('UPDATE SnackBatches SET QuantityRemaining = 0 WHERE SnackBatchId = ?', [batch.SnackBatchId]);
        }
    }
    updateSnackPriceAndStock(snackId);
}
